File: max/engine/ngram_engine.js
"use strict";

/**
 * ngram_engine.js — variable-order chord n-gram with a small stateful history.
 *
 * Direct CommonJS port of
 * versions/v4-theory/python/src/engines/ngram_engine.py (class NgramEngine).
 *
 * The learned JSON model supplies the probability mass; this engine remembers
 * the last few chords, backs off from the longest matching context to shorter
 * ones (and finally to the global unigram table), reshapes counts by a
 * temperature, and samples an output. Parity with the Python reference is
 * asserted in ngram_engine.test.js against golden fixtures.
 *
 * Load-bearing Python behaviors reproduced verbatim:
 *   - appendHistory dedupes only *consecutive* repeats, then trims to the last
 *     max_order chords (Python `del self._history[:-max_order]`).
 *   - lookupCounts keys are `str(order)` in orders and the context joined by
 *     "|"; an empty counts list is treated as a miss (Python `if found:`).
 *   - distribution raises each count to the power 1/temperature (counts are
 *     floored at 0 first) and normalizes; a non-positive total yields [].
 *   - The model must declare version === 1, an object `orders`, and a truthy
 *     `global`; anything else is rejected exactly like Python.
 *
 * The model's `run_length_encoded` flag is metadata only: orders/global are
 * stored as plain [chord, count] pairs and are read as such (no RLE decode).
 *
 * The pure helpers (appendHistory, lookupCounts, distribution) are exported so
 * the parity tests can assert them directly against dumped Python I/O.
 */

const fs = require("fs");
const { makeRng } = require("./rng.js");
const { reduceForNeural, parseChord } = require("./theory.js");

const NAME = "ngram";

class NgramModelError extends Error {
  constructor(message) {
    super(message);
    this.name = "NgramModelError";
  }
}

function isPlainObject(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

// ---- pure helpers (mirror the Python private methods) ----------------------

// _append: dedupe consecutive repeats, then keep only the last maxOrder chords.
// Returns a NEW history array; does not mutate the input.
function appendHistory(history, chord, maxOrder) {
  const next = history.slice();
  if (next.length === 0 || next[next.length - 1] !== chord) {
    next.push(chord);
  }
  // Python: del self._history[:-self._max_order] -> keep the final maxOrder.
  return next.length > maxOrder ? next.slice(-maxOrder) : next;
}

// _lookup: longest-context backoff, then the global table.
function lookupCounts(orders, globalCounts, history, maxOrder) {
  const maxN = Math.min(maxOrder, history.length);
  for (let order = maxN; order >= 1; order--) {
    const key = history.slice(-order).join("|");
    const table = orders[String(order)];
    const found = table ? table[key] : undefined;
    if (found && found.length) {
      return found;
    }
  }
  return globalCounts;
}

// _distribution: count^(1/tau) then normalize; [] on total <= 0.
function distribution(counts, temperature) {
  const invTau = 1.0 / temperature;
  const weighted = counts.map((pair) => [
    String(pair[0]),
    Math.pow(Math.max(Number(pair[1]), 0.0), invTau),
  ]);
  let total = 0;
  for (const w of weighted) {
    total += w[1];
  }
  if (total <= 0) {
    return [];
  }
  return weighted.map((w) => [w[0], w[1] / total]);
}

/**
 * createNgramEngine({modelPath, fallback, rng, temperature}) -> engine.
 *
 *   modelPath   path to the n-gram JSON model (loaded and validated eagerly).
 *   fallback    "echo_input" (default) | "error_only" | "global_top" |
 *               "random_source" — matches the Python fallback modes.
 *   rng         an rng function () => float in [0,1), OR a numeric seed passed
 *               to rng.js makeRng(), OR null/undefined for Math.random.
 *   temperature sampling temperature (clamped to a 0.05 floor). Default 1.0.
 */
function createNgramEngine(options) {
  const opts = options || {};
  const { modelPath } = opts;
  const fallback = opts.fallback === undefined ? "echo_input" : opts.fallback;
  const temperatureOpt = opts.temperature === undefined ? 1.0 : opts.temperature;
  const rngFn = typeof opts.rng === "function" ? opts.rng : makeRng(opts.rng);

  let model;
  try {
    model = JSON.parse(fs.readFileSync(modelPath, "utf8"));
  } catch (exc) {
    throw new NgramModelError(`cannot load n-gram model ${modelPath}: ${exc.message}`);
  }

  if (model.version !== 1) {
    throw new NgramModelError(`unsupported n-gram schema version: ${JSON.stringify(model.version)}`);
  }
  const globalTable = model.global;
  // Python's `not model.get("global")` rejects None/[]/{}/"" alike, and the
  // engine indexes global as a list (self._global[0]); so require a non-empty
  // ARRAY. (A plain {} is truthy in JS but empty-and-non-indexable — Python
  // rejects it and so must we.)
  if (!isPlainObject(model.orders) || !Array.isArray(globalTable) || globalTable.length === 0) {
    throw new NgramModelError(`n-gram model ${modelPath} is missing orders/global`);
  }

  const orders = model.orders;
  const globalCounts = globalTable;
  const maxOrder = Math.max(1, Math.trunc(Number(model.max_order === undefined ? 4 : model.max_order)));

  let temperature = Math.max(0.05, Number(temperatureOpt));
  let history = [];

  function reset() {
    history = [];
  }

  function setTemperature(t) {
    temperature = Math.max(0.05, Number(t));
  }

  // _choose: weighted pick. Python's random.Random cannot be reproduced, so
  // only distribution/structure is parity-tested — never the individual draw.
  function choose(choices) {
    const symbols = choices.map((c) => c[0]);
    const weights = choices.map((c) => c[1]);
    let totalWeight = 0;
    for (const w of weights) {
      totalWeight += w;
    }
    let target = rngFn() * totalWeight;
    let picked = symbols[symbols.length - 1];
    for (let i = 0; i < choices.length; i++) {
      target -= weights[i];
      if (target < 0) {
        picked = symbols[i];
        break;
      }
    }
    return [picked, choices[symbols.indexOf(picked)][1]];
  }

  function applyFallback(chord) {
    const error = `unknown chord: ${chord}`;
    if (fallback === "error_only") {
      return { output: null, probability: null, candidates: 0, fallbackUsed: true, error };
    }
    if (fallback === "global_top" && globalCounts.length) {
      return { output: String(globalCounts[0][0]), probability: null, candidates: 0, fallbackUsed: true, error };
    }
    if (fallback === "random_source") {
      const sources = Object.keys(orders["1"] || {});
      if (sources.length) {
        const idx = Math.floor(rngFn() * sources.length);
        return { output: sources[idx], probability: null, candidates: sources.length, fallbackUsed: true, error };
      }
    }
    return { output: chord || null, probability: null, candidates: 0, fallbackUsed: true, error };
  }

  function sample(rawInput, sampleOpts) {
    const o = sampleOpts || {};
    const session = o.session === undefined ? true : o.session;
    const candidateSelector = o.candidateSelector;

    const trimmed = String(rawInput).trim();
    const source = reduceForNeural(trimmed);
    if (!source || parseChord(source) === null) {
      return applyFallback(trimmed);
    }
    if (!session) {
      reset();
    }
    history = appendHistory(history, source, maxOrder);
    const choices = distribution(lookupCounts(orders, globalCounts, history, maxOrder), temperature);
    if (!choices.length) {
      return applyFallback(source);
    }
    let selected;
    try {
      selected = candidateSelector
        ? candidateSelector(source, choices, NAME)
        : choose(choices);
    } catch (exc) {
      return { output: null, probability: null, candidates: 0, fallbackUsed: true, error: String(exc && exc.message !== undefined ? exc.message : exc) };
    }
    const output = selected[0];
    const probability = selected[1];
    history = appendHistory(history, output, maxOrder);
    return { output, probability, candidates: choices.length, fallbackUsed: false, error: null };
  }

  return {
    name: NAME,
    sample,
    resetSession: reset,
    setTemperature,
    get history() {
      return history.slice();
    },
  };
}

module.exports = {
  createNgramEngine,
  NgramModelError,
  appendHistory,
  lookupCounts,
  distribution,
};

File: versions/v4-theory/python/src/engines/ngram_engine.py
"""Variable-order chord n-gram trained from the local JazzNet sequences."""

from __future__ import annotations

import json
import math
import random
from pathlib import Path
from typing import Callable

from ..harmony.theory import parse_chord, reduce_for_neural
from .base import SampleResult

CandidateSelector = Callable[[str, list[tuple[str, float]], str], tuple[str, float]]


class NgramModelError(RuntimeError):
    pass


class NgramEngine:
    """Back-off n-gram with a small stateful chord history.

    This complements Markov order 1 and recurrent neural models: it remembers
    exact 2–4 chord cells, making turnarounds and ii–V–I patterns more stable,
    while remaining inspectable and fast enough to run at Max startup.
    """

    name = "ngram"

    def __init__(
        self,
        model_path: str | Path,
        *,
        fallback: str = "echo_input",
        seed: int | None = None,
        temperature: float = 1.0,
    ) -> None:
        path = Path(model_path)
        try:
            model = json.loads(path.read_text())
        except (OSError, ValueError) as exc:
            raise NgramModelError(f"cannot load n-gram model {path}: {exc}") from exc
        if model.get("version") != 1:
            raise NgramModelError(
                f"unsupported n-gram schema version: {model.get('version')!r}"
            )
        if not isinstance(model.get("orders"), dict) or not model.get("global"):
            raise NgramModelError(f"n-gram model {path} is missing orders/global")
        self._orders: dict[str, dict[str, list[list]]] = model["orders"]
        self._global: list[list] = model["global"]
        self._max_order = max(1, int(model.get("max_order", 4)))
        self._fallback = fallback
        self._rng = random.Random(seed)
        self._temperature = max(0.05, float(temperature))
        self._history: list[str] = []

    @property
    def history(self) -> tuple[str, ...]:
        return tuple(self._history)

    def reset_session(self) -> None:
        self._history.clear()

    def set_temperature(self, temperature: float) -> None:
        self._temperature = max(0.05, float(temperature))

    def _append(self, chord: str) -> None:
        if not self._history or self._history[-1] != chord:
            self._history.append(chord)
        del self._history[:-self._max_order]

    def _lookup(self) -> list[list]:
        max_n = min(self._max_order, len(self._history))
        for order in range(max_n, 0, -1):
            key = "|".join(self._history[-order:])
            found = self._orders.get(str(order), {}).get(key)
            if found:
                return found
        return self._global

    def _distribution(self, counts: list[list]) -> list[tuple[str, float]]:
        inv_tau = 1.0 / self._temperature
        weighted = [(str(chord), math.pow(max(float(count), 0.0), inv_tau)) for chord, count in counts]
        total = sum(weight for _, weight in weighted)
        if total <= 0:
            return []
        return [(chord, weight / total) for chord, weight in weighted]

    def _choose(self, choices: list[tuple[str, float]]) -> tuple[str, float]:
        symbols = [symbol for symbol, _ in choices]
        weights = [probability for _, probability in choices]
        symbol = self._rng.choices(symbols, weights=weights, k=1)[0]
        return symbol, choices[symbols.index(symbol)][1]

    def sample(
        self,
        raw_input: str,
        *,
        session: bool = True,
        candidate_selector: CandidateSelector | None = None,
    ) -> SampleResult:
        source = reduce_for_neural(raw_input.strip())
        if not source or parse_chord(source) is None:
            return self._apply_fallback(raw_input.strip())
        if not session:
            self.reset_session()
        self._append(source)
        choices = self._distribution(self._lookup())
        if not choices:
            return self._apply_fallback(source)
        try:
            selected = (
                candidate_selector(source, choices, self.name)
                if candidate_selector is not None
                else self._choose(choices)
            )
        except ValueError as exc:
            return SampleResult(None, None, 0, True, error=str(exc))
        output, probability = selected
        self._append(output)
        return SampleResult(output, probability, len(choices), False)

    def _apply_fallback(self, chord: str) -> SampleResult:
        error = f"unknown chord: {chord}"
        if self._fallback == "error_only":
            return SampleResult(None, None, 0, True, error=error)
        if self._fallback == "global_top" and self._global:
            return SampleResult(str(self._global[0][0]), None, 0, True, error=error)
        if self._fallback == "random_source":
            sources = list(self._orders.get("1", {}))
            if sources:
                return SampleResult(self._rng.choice(sources), None, len(sources), True, error=error)
        return SampleResult(chord or None, None, 0, True, error=error)
